core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala

```diff
@@ -74,19 +80,27 @@ private[history] class HistoryPage(parent: HistoryServer) extends WebUIPage("")
             // page, `...` will be displayed.
             if (allAppsSize > 0) {
               val leftSideIndices =
-                rangeIndices(actualPage - plusOrMinus until actualPage, 1 < _, requestedIncomplete)
+                rangeIndices(actualPage - plusOrMinus until actualPage, 1 < _, requestedIncomplete,
+                  requestedQuery)
               val rightSideIndices =
                 rangeIndices(actualPage + 1 to actualPage + plusOrMinus, _ < pageCount,
-                  requestedIncomplete)
+                  requestedIncomplete, requestedQuery)
 
               <h4>
+                <form action="" method="GET">
+                  Search: <input type="text" name="query" value={requestedQuery} />
```

Comment:
It's difficult for users to know the targets searched by keywords. How about set initial message "Input App Names" or like that to the text field and remove "Search" label.
It's just one idea. Someone might have better idea.